### PR TITLE
ARROW-9465: [Python] Improve ergonomics of compute module

### PR DESCRIPTION
**Pull request by pitrou**
Lots of assorted things here:
* Automatically generate global wrappers for calling
  registered compute functions
* Improve metadata of such wrappers (e.g. name, docstring)
* Make it easier to pass options (for example via kwargs)
* Type-check options
* Add some docstrings
* Expose more function attributes (e.g. arity)
* Fix some crashes

### CODE REVIEW

File: python/pyarrow/_compute.pyx [hunk 1]
```diff
@@ -140,6 +162,29 @@ cdef class ScalarAggregateKernel(Kernel):
...
+      string predicates...
+
+    * "vector" functions apply a collection-wise computation, such that
+      each item in the output may depend on the values of several items
+      in each input.  Examples: dictionary encoding, sorting, extracting
+      unique values...
+
+    * "aggregate" functions reduce the dimensionality of the inputs by
+      applying a reduction function.  Examples: sum, minmax, mode...
+
```
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:186]: Should "meta" be explained here as well?

File: cpp/src/arrow/compute/api_aggregate.h [hunk 2]
```diff
@@ -66,7 +66,7 @@ struct ARROW_EXPORT MinMaxOptions : public FunctionOptions {
     /// Skip null values
     SKIP = 0,
     /// Any nulls will result in null output
-    OUTPUT_NULL
+    EMIT_NULL
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/api_aggregate.h:69]: Can we simply change this? (is it public on the C++ side?) And it is changed for consistency with how it's called elsewhere?

**Review by jorisvandenbossche — COMMENTED** [review]
Nice improvements!

File: python/pyarrow/_compute.pyx (hunk 1 above)
Comment (pitrou) [reply python/pyarrow/_compute.pyx:186]: Yes, you're right.

File: cpp/src/arrow/compute/api_aggregate.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/api_aggregate.h:69]: The next version will be 2.0.0, meaning it's ok to break APIs. Of course, we should try to minimize breakage, but this one aims to fix an inconsistency.

**Comment by pitrou** [comment]
CI failures here look unrelated (though someone should really take a look at them).

**Pull request closed without merging**